sandbox.go

```diff
@@ -0,0 +1,488 @@
+		if err := s.updateGateway(ep); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *sandboxData) clearNetworkResources(ep *endpoint) error {
+	sb := s.sandbox()
+	for _, i := range sb.Info().Interfaces() {
+		// Only remove the interfaces owned by this endpoint from the sandbox.
+		if ep.hasInterface(i.SrcName()) {
+			if err := i.Remove(); err != nil {
+				logrus.Debugf("Remove interface failed: %v", err)
+			}
+		}
+	}
+
+	ep.Lock()
+	joinInfo := ep.joinInfo
+	ep.Unlock()
+
+	// Remove non-interface routes.
+	for _, r := range joinInfo.StaticRoutes {
+		if err := sb.RemoveStaticRoute(r); err != nil {
+			logrus.Debugf("Remove route failed: %v", err)
+		}
+	}
+
+	s.Lock()
+	if len(s.endpoints) == 0 {
+		// s.endpoints should never be empty and this is unexpected error condition
+		// We log an error message to note this down for debugging purposes.
+		logrus.Errorf("No endpoints in sandbox while trying to remove endpoint %s", ep.Name())
+		s.Unlock()
+		return nil
+	}
+
+	highEpBefore := s.endpoints[0]
+	var (
+		i int
+		e *endpoint
+	)
+	for i, e = range s.endpoints {
+		if e == ep {
+			break
+		}
+	}
+	heap.Remove(&s.endpoints, i)
+	var highEpAfter *endpoint
+	if len(s.endpoints) > 0 {
+		highEpAfter = s.endpoints[0]
+	}
+	delete(s.epPriority, ep.ID())
+	s.Unlock()
+
+	if highEpBefore != highEpAfter {
+		s.updateGateway(highEpAfter)
+	}
+
+	return nil
+}
+
+func (s *sandboxData) Delete() error {
+	s.Lock()
+	c := s.controller
+	id := s.id
+	eps := make([]*endpoint, len(s.endpoints))
+	for i, ep := range s.endpoints {
+		eps[i] = ep
+	}
+	s.Unlock()
+
+	// Detach from all containers
+	for _, ep := range eps {
+		if err := ep.Leave(s); err != nil {
+			logrus.Warnf("Failed detaching sandbox %s from endpoint %s: %v\n", s.ID(), ep.ID(), err)
+		}
+	}
+
+	sbox := s.sandbox()
+	if sbox != nil {
+		sbox.Destroy()
+	}
+
+	c.Lock()
+	delete(c.containers, id)
+	delete(c.sandboxes, s.data.SandboxKey)
+	c.Unlock()
+
+	return nil
+}
+
+func (s *sandboxData) Info() SandboxInfo {
+	return s
+}
+
+func (s *sandboxData) Statistics() (map[string]*sandbox.InterfaceStatistics, error) {
+	m := make(map[string]*sandbox.InterfaceStatistics)
+
+	sbox := s.sandbox()
+	if sbox == nil {
+		return m, nil
+	}
+
+	var err error
+	for _, i := range sbox.Info().Interfaces() {
+		if m[i.DstName()], err = i.Statistics(); err != nil {
+			return m, err
+		}
+	}
+
+	return m, nil
+}
+
+// NewSandbox cretes a new sandbox for the passed container id
+func (c *controller) NewSandbox(containerID string, options ...SandboxOption) (Sandbox, error) {
+	if containerID == "" {
+		return nil, types.BadRequestErrorf("invalid container ID")
+	}
+
+	c.Lock()
+	currSData, present := c.containers[containerID]
+	c.Unlock()
+	if present {
+		return nil, fmt.Errorf("container %s is already present: %v", containerID, currSData)
+	}
+
+	// Create sData and process options first. Key generation depends on an option
+	sData := &sandboxData{
+		id:         containerID,
```

Comment:
As discussed, IMO id should be an UUID managed by libnetwork, whereas containerID is a data that sandboxData must store for its operation.